#!/usr/bin/env bash

FASTQC=$(grep -i 'fastqc' $1 | sort -u |awk '{print $2}' |sed -E 's/v//')
MULTIQC=$(grep -i 'multiqc' $1 | sort -u |awk '{print $3}' |sed -E 's/v//')
DADA=$(grep -i 'dada2' $1 | sort -u |awk '{print $2}' |sed -E 's/v//')
DECIPHER=$(grep -i 'decipher' $1 | sort -u | awk '{print $2}' |sed -E 's/v//')
CUTADAPT=$(grep -i 'cutadapt' $1 | sort -u |awk '{print $2}' |sed -E 's/v//')
ANCOMBC=$(grep -i 'ancombc' $1 | sort -u |awk '{print $2}' |sed -E 's/v//')
DESEQ=$(grep -i 'deseq' $1 | sort -u |awk '{print $2}' |sed -E 's/v//')
PHYLOSEQ=$(grep -i 'phyloseq' $1 | sort -u |awk '{print $2}' |sed -E 's/v//')
VEGAN=$(grep -i 'vegan' $1 | sort -u |awk '{print $2}' |sed -E 's/v//')
R=$(grep -i '^r ' "$1" | sort -u | awk '{print $3}' | sed -E 's/v//')
GGPLOT=$(grep -i 'ggplot2' $1 | sort -u |awk '{print $2}' |sed -E 's/v//')


PROTOCOL_ID=$2 # GL-DPPD-7104-C

if [[ -s "$3" ]]; then
	RAREFACTION_DEPTH=$(cat $3)
else
	RAREFACTION_DEPTH="N/A"
fi

PROTOCOL="Data were processed as described in ${PROTOCOL_ID} (https://github.com/nasa/GeneLab_Data_Processing/blob/master/Amplicon/Illumina/Pipeline_GL-DPPD-7104_Versions/${PROTOCOL_ID}.md), using workflow NF_AmpIllumina v1.0.4 (https://github.com/nasa/GeneLab_AmpliconSeq_Workflow/blob/v1.0.4/README.md).

Quality control: Primers were removed from raw reads using cutadapt v${CUTADAPT}. Quality assessment of raw and trimmed reads was performed with FastQC v${FASTQC} and reports were summarized with MultiQC v${MULTIQC}.

ASV table generation and taxonomy assignment: DADA2 v${DADA} was utilized for quality trimming, filtering, and inference of amplicon sequence variants (ASVs) with the following parameters: truncLen=c(left_truncLen, right_trucLen), maxEE=c(left_maxEE, right_maxEE), and truncQ=2. Taxonomy assignment of the ASVs was performed with DECIPHER v${DECIPHER} against the SILVA r138_2 reference database. Sample-wise and group-wise relative abundance was calculated and plotted for each taxon level.

Diversity analysis: For alpha (within samples) and beta (between samples) analyses, the ASV count table generated by DADA2 were normalized using two different methods: variance stabilizing transformation (VST) with DESeq2 v${DESEQ} and rarefaction to even depth with phyloseq v${PHYLOSEQ}. For normalization using rarefaction, samples were rarefied to even depth at ${RAREFACTION_DEPTH} sequences per sample. To estimate alpha-diversity, the total number of observed features, Chao1, Simpson, and Shannon’s diversity index (H′) and rarefaction curves were calculated using vegan v${VEGAN}. For beta diversity, Bray-Curtis and Euclidean distance matrices were generated using vegan v${VEGAN} on the rarefied and VST normalized ASV tables, respectively. 

Statistical analysis: Statistical analyses and graphical displays were conducted and generated using R v${R} and ggplot2 v${GGPLOT}. Alpha diversity: The non-parametric Wilcoxon rank-sum or Dunn test was used to detect significant differences between two or multiple groups, respectively. Beta diversity: Principal co-ordinate analysis (PCoA) was performed on both Bray Curtis and Euclidean distances between samples. The principal co-ordinates were visualized as scatter plots and distances were visualized as dendrograms using ggplot2. The non-parametric permutational multivariate analysis of variance (PERMANOVA) was used to determine the overall changes in microbial communities between samples and group clustering with 999 permutations based on both distance matrices using the adonis2() function in vegan package v${VEGAN}.

Differential abundance testing: Differential abundance testing was performed using three different methods: ANCOM-BC v${ANCOMBC}, ANCOM-BC2 v${ANCOMBC} and DESeq2 v${DESEQ}. ANCOM-BC was run once for each pair-wise group comparison using a Z-test to generate p-values derived from the W statistic in a log-linear regression model with bias correction. ANCOM-BC2 utilized a pairwise directional test to perform pair-wise group comparisons. DESeq2 was used to perform pair-wise group comparisons using the Wald test with a negative binomial generalized linear model. FDR was used to adjust for multiple comparisons testing and to generate an adjusted p-value in all 3 methods. To visualize group-wise differences, volcano plots were generated for each pair-wise comparison for each differential abundance method. Only ASVs with an adjusted p-value <= 0.1 were considered significant." 

printf "%s\n" "$PROTOCOL"
